### PR TITLE
Completed homework (Oleksii Hashparych)

### PR DESCRIPTION
**Pull request by lesykgg**

### CODE REVIEW

File: lib/topics/state_and_behaviour.rb [hunk 1]
```diff
@@ -2,4 +2,26 @@
...
+      @model = args[:model] || 'Mercedes W210'
+      @current_speed = 0
+    end
+
+    def self.default_car
+      new
+    end
+
+    def speed_up(increase_speed)
+      @current_speed += increase_speed
```
Comment (vbyno) [inline lib/topics/state_and_behaviour.rb]: As Alex said during the lecture, if the setter is defined, it should be used in all assignments

File: lib/topics/state_and_behaviour.rb [hunk 2]
```diff
@@ -2,4 +2,26 @@
 
 # exercise state and behaviour
 module StateAndBehaviour
+  # Describes car's state and behaviour
+  class Car
+    attr_accessor :year, :color, :model, :current_speed
+    def initialize(args)
```
Comment (vbyno) [inline lib/topics/state_and_behaviour.rb]: An empty line is missed

File: lib/topics/state_and_behaviour.rb [hunk 3]
```diff
@@ -2,4 +2,26 @@
...
+      @year = args[:year]   || 2000
```
Comment (vbyno) [inline lib/topics/state_and_behaviour.rb]: maybe, these are attributes for `default_car` method rather than `initializer`?

Comment (lesykgg) [inline lib/topics/state_and_behaviour.rb]: Car::#initialize#test_0003_provides default values
as was described at test case

Comment (lesykgg) [reply lib/topics/state_and_behaviour.rb]: @vbyno you're about args, or about default parameter for `@year`?

File: lib/topics/state_and_behaviour.rb [hunk 4]
```diff
@@ -2,4 +2,29 @@
...
+      @model = args[:model] || 'Mercedes W210'
+      @current_speed = 0
+    end
+
+    def self.default_car
+      new
+    end
+
+    def speed_up(increase_speed)
+      self.current_speed = current_speed + increase_speed
```
Comment (vbyno) [inline lib/topics/state_and_behaviour.rb]: `+=` works here as well

File: lib/topics/state_and_behaviour.rb [hunk 5]
```diff
@@ -2,4 +2,29 @@
...
+    attr_accessor :year, :color, :model, :current_speed
+
+    def initialize(args)
+      @year = args[:year]   || 2000
+      @color = args[:color] || 'White'
+      @model = args[:model] || 'Mercedes W210'
+      @current_speed = 0
+    end
+
+    def self.default_car
```
Comment (vbyno) [inline lib/topics/state_and_behaviour.rb]: What about the [class structure](https://github.com/bbatsov/ruby-style-guide#consistent-classes)? Class methods have to be defined before instance methods

File: lib/topics/state_and_behaviour.rb [hunk 6]
```diff
@@ -2,4 +2,27 @@
...
+  # Describes car's state and behaviour
+  class Car
+    attr_accessor :year, :color, :model, :current_speed
+
+    def self.default_car
+      new
+    end
+
+    def initialize(args)
+      @year = args.fetch(:year, 2000)
```
Comment (vbyno) [inline lib/topics/state_and_behaviour.rb]: @lesykgg , have you taken a look at [keyword arguments](https://robots.thoughtbot.com/ruby-2-keyword-arguments) in Ruby?

File: lib/topics/state_and_behaviour.rb [hunk 7]
```diff
@@ -2,4 +2,27 @@
 
 # exercise state and behaviour
 module StateAndBehaviour
+  # Describes car's state and behaviour
+  class Car
+    attr_accessor :year, :color, :model, :current_speed
```
Comment (vbyno) [inline lib/topics/state_and_behaviour.rb]: Do we need public attribute writers? Or just getters?

File: lib/topics/state_and_behaviour.rb (hunk 6 above)
Comment (lesykgg) [reply lib/topics/state_and_behaviour.rb]: @vbyno, yes I have. But tests says that initialize should take an hash of options

**Comment by lesykgg** [comment]
@vbyno, fixed. Or we can use instance variables in `speed_up` & `push_break` instead setters?

File: lib/topics/state_and_behaviour.rb [hunk 8]
```diff
@@ -2,4 +2,28 @@
...
+
+    def initialize(args)
+      @year = args.fetch(:year, 2000)
+      @color = args.fetch(:color, 'Silver')
+      @model = args.fetch(:model, 'Mercedes-Benz W210')
+      @current_speed = 0
+    end
+
+    def speed_up(increase_speed)
+      self.current_speed += increase_speed
```
Comment (sveredyuk) [inline lib/topics/state_and_behaviour.rb]: @current_speed instead .self

**Comment by lesykgg** [comment]
@sveredyuk, fixed

**Review by sveredyuk — APPROVED** [review]

File: lib/topics/state_and_behaviour.rb [hunk 9]
```diff
@@ -2,4 +2,27 @@
 
 # exercise state and behaviour
 module StateAndBehaviour
+  # Describes car's state and behaviour
+  class Car
+    attr_reader :year, :color, :model, :current_speed
+
+    def self.default_car
+      new
```
Comment (sveredyuk) [inline lib/topics/state_and_behaviour.rb]: new will fails with Argument error

File: lib/topics/state_and_behaviour.rb [hunk 10]
```diff
@@ -2,4 +2,27 @@
...
+    end
+
+    def initialize(args)
```
Comment (sveredyuk) [inline lib/topics/state_and_behaviour.rb]: args = {} # will not fail new

**Comment by lesykgg** [comment]
@sveredyuk, fixed

**Pull request closed without merging**